### PR TITLE
[MM-52828] Migrate "components/admin_console/team_channel_settings/group/index.js" to typescript

### PR DESCRIPTION
**Pull request by ridwankabeer435**
#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/23432
https://mattermost.atlassian.net/browse/MM-52828

#### Release Note

```release-note

```

### CODE REVIEW

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import {ActionCreatorsMapObject, bindActionCreators, Dispatch} from 'redux';
+import {ActionFunc} from 'mattermost-redux/types/actions';
+import {getGroups} from 'mattermost-redux/actions/groups';
+import {t} from 'utils/i18n';
+import {Group} from '@mattermost/types/groups';
+import List from './group_list';
+import {GlobalState} from 'types/store';
+
+type Actions = {
+
```
Comment (sinansonmez) [inline webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:14]: nit: no extra lines please

File: webapp/channels/src/components/admin_console/team_channel_settings/channel/details/channel_groups.tsx [hunk 2]
```diff
@@ -62,7 +62,7 @@ export const ChannelGroups: React.FunctionComponent<ChannelGroupsProps> = (props
             {channel.id && (
                 <GroupList
                     channel={channel}
-                    groups={groups}
+                    groups={groups as Group[]}
```
Comment (sinansonmez) [inline webapp/channels/src/components/admin_console/team_channel_settings/channel/details/channel_groups.tsx:65]: Why do we need a type casting here?

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        removeGroup: ownProps.onGroupRemoved,
+        setNewGroupRole: ownProps.setNewGroupRole,
+    };
+}
+
+function mapDispatchToProps(dispatch: Dispatch) {
+    return {
+        actions: bindActionCreators<ActionCreatorsMapObject<ActionFunc>, Actions>({
+            getData: () => getGroups(),
+
```
Comment (sinansonmez) [inline webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:40]: nit: no extra line please

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        emptyListTextDefaultMessage: ownProps.isModeSync ? 'At least one group must be specified' : 'No groups specified yet',
+        removeGroup: ownProps.onGroupRemoved,
+        setNewGroupRole: ownProps.setNewGroupRole,
+    };
+}
+
+function mapDispatchToProps(dispatch: Dispatch) {
+    return {
+        actions: bindActionCreators<ActionCreatorsMapObject<ActionFunc>, Actions>({
+            getData: () => getGroups(),
```
Comment (sinansonmez) [inline webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:39]: `getGroups` was not used in the older version. Why do we need to use it here or am I missing something?

**Review by sinansonmez — CHANGES_REQUESTED** [review]
Thanks for the effort. I just have a few small comments

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts (hunk 4 above)
Comment (ridwankabeer435) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:39]: I was getting compilation errors with the older version of the function. I guessed that using  `getGroups` action would give us the same result as `Promise.resolve()`. Do you still want to retain the older version of this function?

File: webapp/channels/src/components/admin_console/team_channel_settings/channel/details/channel_groups.tsx (hunk 2 above)
Comment (ridwankabeer435) [reply webapp/channels/src/components/admin_console/team_channel_settings/channel/details/channel_groups.tsx:65]: I kept getting the following error without the type casting:

`Type 'Partial<Group>[]' is not assignable to type 'Group[]'.
  Type 'Partial<Group>' is not assignable to type 'Group'.
    Types of property 'id' are incompatible.
      Type 'string | undefined' is not assignable to type 'string'.
        Type 'undefined' is not assignable to type 'string'.ts(2322)`

Comment (sinansonmez) [reply webapp/channels/src/components/admin_console/team_channel_settings/channel/details/channel_groups.tsx:65]: It seems to me id is creating an issue because it can be also undefined. GroupList doesnt like that. We can make sure that id is not undefined by assigning it to empty string when it is undefined. Alternatively, we can change the type of groups as `Partial<Group>[]` in GroupList.

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts (hunk 4 above)
Comment (sinansonmez) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:38]: I dont think `getGroups` is the correct action to be used here. Because if you check `GroupList` action is passed down to `AbstractList` component and `getGroup` type definition is different there. 

In the meantime I am also wondering how this component is working properly with  `Promise.resolve()`. 

Therefore I would request another review from the internal member who knows this area of the code base cc: @M-ZubairAhmed

File: webapp/channels/src/components/admin_console/team_channel_settings/channel/details/channel_groups.tsx (hunk 2 above)
Comment (ridwankabeer435) [reply webapp/channels/src/components/admin_console/team_channel_settings/channel/details/channel_groups.tsx:65]: @sinansonmez I tried changing the type of groups to `Partial<Group>[]` but then `AbstrasctList` inside `GroupList` does not like that because of its `data: Array<Group | TeamWithMembership>`. I then tried changing that to `data: Array<Partial<Group> | Partial<TeamWithMembership>>`, but then TypeScript's transpiler complains at `return this.props.data.slice(offset, offset + PAGE_SIZE).map(this.props.renderRow);` inside the renderRows functions, which expects a type of `JSX.Element | JSX.Element[]`

This is the error I get:
`Argument of type '(item: Group | TeamWithMembership) => Element' is not assignable to parameter of type '(value: Partial<Group> | Partial<TeamWithMembership>, index: number, array: (Partial<Group> | Partial<TeamWithMembership>)[]) => Element'.
  Types of parameters 'item' and 'value' are incompatible.
    Type 'Partial<Group> | Partial<TeamWithMembership>' is not assignable to type 'Group | TeamWithMembership'.
      Type 'Partial<Group>' is not assignable to type 'Group | TeamWithMembership'.
        Type 'Partial<Group>' is not assignable to type 'Group'.
          Types of property 'id' are incompatible.
            Type 'string | undefined' is not assignable to type 'string'.
              Type 'undefined' is not assignable to type 'string'.`

Comment (ridwankabeer435) [reply webapp/channels/src/components/admin_console/team_channel_settings/channel/details/channel_groups.tsx]: I have updated `channel_groups.tsx` to set default values for each missing `Partial<Group>` attribute before passing the `groups` prop to `GroupList`. However, I still remain uncertain about the default values. Would you suggest making each attribute in the `Group` type optional?

Comment (sinansonmez) [reply webapp/channels/src/components/admin_console/team_channel_settings/channel/details/channel_groups.tsx]: yes, we can keep the default values in the test

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts (hunk 4 above)
Comment (sinansonmez) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:38]: @M-ZubairAhmed any idea here?

Comment (ridwankabeer435) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:38]: @M-ZubairAhmed, would really appreciate your insights about the most suitable action for getting groups. This is my first time contributing to the Mattermost codebase

Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:38]: Oops sorry i dont know how i missed this notification for so long. Let me answer you tomorrow

**Comment by M-ZubairAhmed** [comment]
@ridwankabeer435 can you please update your fork to `mattermost`

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts (hunk 4 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:38]: lets use the same promise.resolve, i think the component should have been properly added the fallback when getData was not defined.

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:1]: Lets follow `webapp/channels/src/components/team_controller/index.ts` to implement PropsFromRedux approach here

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts [hunk 5]
```diff
@@ -32,13 +26,12 @@ function mapStateToProps(state: GlobalState, ownProps: OwnProps) {
...
-    };
-}
+const mapDispatchToProps = {
+    getData: () => Promise.resolve(),
+};
+
+const connector = connect(mapStateToProps, mapDispatchToProps);
 
-export default connect(mapStateToProps, mapDispatchToProps)(List);
+export type PropsFromRedux = ConnectedProps<typeof connector>;
```
Comment (ridwankabeer435) [inline webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:36]: @M-ZubairAhmed is this what you had in mind? I am getting the following error in `channel_groups.tsx` and `team_groups.tsx` after making the changes: `Property 'actions' is missing in type '{ team: Team; isModeSync: boolean; groups: Group[]; totalGroups: number; onGroupRemoved: (groupId: string) => void; setNewGroupRole: (groupId: string) => void; type: string; isDisabled: boolean | undefined; }' but required in type 'Omit<ClassAttributes<GroupList> & Props, "data" | "total" | "emptyListTextId" | "emptyListTextDefaultMessage" | "removeGroup" | "setNewGroupRole">'`;

basically `GroupList` is now expecting its parent components to pass an implementation of `getData: () => Promise<Group[]>` as a prop value to `GroupList` 's `actions` prop.

Comment (ridwankabeer435) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:36]: Please do not run any integration and end-to-end test at this point.

File: webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx [hunk 6]
```diff
@@ -40,7 +40,7 @@ const Header = () => {
     );
 };
 
-interface Props {
+type Props = {
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx:43]: [2]

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.js [hunk 7]
```diff
@@ -1,30 +0,0 @@
...
-    };
-}
-
-function mapDispatchToProps() {
-    return {
-        actions: {
-            getData: () => Promise.resolve(),
-        },
-    };
-}
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/admin_console/team_channel_settings/group/index.js:27]: [3]

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts (hunk 5 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:36]: @ridwankabeer435 i have pulled this branch locally and here are my observations/suggestions

- As seen in https://github.com/mattermost/mattermost/pull/23511#discussion_r1245502729 Lets export this ownProps as well.
- In https://github.com/mattermost/mattermost/pull/23511#discussion_r1245503014 so essentially you Props will just be a combinations of OwnProps and PropsFromRedux
- You might need to add more ownProp types in index file
- Cross check the function return signature of mapDispatchToProps https://github.com/mattermost/mattermost/pull/23511#discussion_r1245504798 the new one should be the same
- Because this is a weird case lets hard type cast the `getData: () => Promise.resolve() as any,` 

I wrote everything here to condense the message clearly in one place

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.js (hunk 7 above)
Comment (ridwankabeer435) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.js:27]: I am not sure what you meant by the [n] comments. Are those potential problem points within the code commits that may be requiring team_groups and channel_groups to pass an action that returns Promise<Group[]>? I am really lost at this point :(

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts [hunk 8]
```diff
@@ -18,16 +20,20 @@ type OwnProps = {
...
         data: ownProps.groups,
+        onPageChangedCallback: ownProps.onPageChangedCallback,
         total: ownProps.totalGroups,
         emptyListTextId: ownProps.isModeSync ? t('admin.team_channel_settings.group_list.no-synced-groups') : t('admin.team_channel_settings.group_list.no-groups'),
         emptyListTextDefaultMessage: ownProps.isModeSync ? 'At least one group must be specified' : 'No groups specified yet',
         removeGroup: ownProps.onGroupRemoved,
         setNewGroupRole: ownProps.setNewGroupRole,
+        actions: {
+            getData: () => Promise.resolve(ownProps.groups),
+        },
```
Comment (ridwankabeer435) [inline webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:31]: Passing `ownProps.groups` to `Promise.resolve` for `getData()` action  pretty much solved most of the issues

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.js (hunk 7 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.js:27]: sorry i just used this a reference points to explain in https://github.com/mattermost/mattermost/pull/23511#discussion_r1245506853
apologies for causing this as confusion. But you are doing really great job 💯

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts [hunk 9]
```diff
@@ -18,16 +20,20 @@ type OwnProps = {
...
     };
 }
 
 const mapDispatchToProps = {
-    getData: () => Promise.resolve(),
+    getData: () => Promise.resolve() as any,
```
Comment (ridwankabeer435) [inline webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:36]: not sure if we need to cast `Promise.resolve()` as any at this point. But I suppose it can work as insurance.

Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:36]: yes this is an exceptional case where even I have no idea why we had this in the first place. But fixing this in this pr is out of scope

File: webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx [hunk 10]
```diff
@@ -40,18 +40,18 @@ const Header = () => {
...
```
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx:43]: this can be just Props = OwnProps & PropsFromRedux

File: webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx [hunk 11]
```diff
@@ -47,7 +47,7 @@ type Props = {
     emptyListTextId: string;
     emptyListTextDefaultMessage: string;
     actions: {
-        getData: () => Promise<Group[]>;
+        getData: () => Promise<any>;
     };
     removeGroup: (gid: string) => void;
     setNewGroupRole: (gid: string) => void;
```
Comment (ridwankabeer435) [inline webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx:53]: kept most of the original props as `group_list_test` starts complaining if I use `type Props = OwbnProps & PropsFromRedux` instead. Not sure if we want to change `group_list_test` to accommodate using `type Props = OwbnProps & PropsFromRedux`.

File: webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts
Comment (ridwankabeer435) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/index.ts:1]: I think the only major change would be to add `actions: {
            getData: () => Promise.resolve(ownProps.groups),
        },` under `mapStateToProps()` before passing the action to `PropsFromRedux`

File: webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx (hunk 11 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx:53]: I would fix type issue in `group_list_test` instead

**Comment by M-ZubairAhmed** [comment]
@ridwankabeer435 do you need help in getting this to finish line?

**Comment by ridwankabeer435** [comment]
> 
I'd like to give it one last shot to fix the issue first. Apologies for leaving the PR hanging for so long

File: webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx (hunk 10 above)
Comment (ridwankabeer435) [reply webapp/channels/src/components/admin_console/team_channel_settings/group/group_list.tsx:43]: Hi @M-ZubairAhmed, I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running `npm test` on the test file): `Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a <Provider>, or pass a custom React context provider to <Provider> and the corresponding React context consumer to Connect(GroupList) in connect options.`

**Comment by M-ZubairAhmed** [comment]
@ridwankabeer435 anything i can do to help you out

**Comment by ridwankabeer435** [comment]
> @ridwankabeer435 do you need help in getting this to finish line?

I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running npm test on the test file): Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a <Provider>, or pass a custom React context provider to <Provider> and the corresponding React context consumer to Connect(GroupList) in connect options.

I am not sure why the transpiler is asking for a `Provider`. Please note I am still fairly new to React with TypeScript. I haven't pushed in the changes because of said errors after modifying the tests

**Comment by M-ZubairAhmed** [comment]
> > @ridwankabeer435 do you need help in getting this to finish line?
> 
> I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running npm test on the test file): Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a , or pass a custom React context provider to and the corresponding React context consumer to Connect(GroupList) in connect options.
> 
> I am not sure why the transpiler is asking for a `Provider`. Please note I am still fairly new to React with TypeScript. I haven't pushed in the changes because of said errors after modifying the tests

Can you push the code so we can verify whats the issue there

**Comment by ridwankabeer435** [comment]
> > > @ridwankabeer435 do you need help in getting this to finish line?
> > 
> > 
> > I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running npm test on the test file): Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a , or pass a custom React context provider to and the corresponding React context consumer to Connect(GroupList) in connect options.
> > I am not sure why the transpiler is asking for a `Provider`. Please note I am still fairly new to React with TypeScript. I haven't pushed in the changes because of said errors after modifying the tests
> 
> Can you push the code so we can verify whats the issue there

I have pushed whatever changes I have made so far. I sincerely apologize for my lack of commitment over the past few month

**Comment by M-ZubairAhmed** [comment]
> > > > @ridwankabeer435 do you need help in getting this to finish line?
> > > 
> > > 
> > > I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running npm test on the test file): Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a , or pass a custom React context provider to and the corresponding React context consumer to Connect(GroupList) in connect options.
> > > I am not sure why the transpiler is asking for a `Provider`. Please note I am still fairly new to React with TypeScript. I haven't pushed in the changes because of said errors after modifying the tests
> > 
> > 
> > Can you push the code so we can verify whats the issue there
> 
> I have pushed whatever changes I have made so far. I sincerely apologize for my lack of commitment over the past few month

Thank you let me have a look but before that can you please the conflicts for the CI to run?

**Comment by ridwankabeer435** [comment]
> > > > > @ridwankabeer435 do you need help in getting this to finish line?
> > > > 
> > > > 
> > > > I updated the line as you suggested. I also included additional props in the index.ts file and updated the group_list_test file as well. However, I keep getting the following error (even after removing the old snapshot file and adding the --updateSnapshot flag while running npm test on the test file): Could not find "store" in the context of "Connect(GroupList)". Either wrap the root component in a , or pass a custom React context provider to and the corresponding React context consumer to Connect(GroupList) in connect options.
> > > > I am not sure why the transpiler is asking for a `Provider`. Please note I am still fairly new to React with TypeScript. I haven't pushed in the changes because of said errors after modifying the tests
> > > 
> > > 
> > > Can you push the code so we can verify whats the issue there
> > 
> > 
> > I have pushed whatever changes I have made so far. I sincerely apologize for my lack of commitment over the past few month
> 
> Thank you let me have a look but before that can you please the conflicts for the CI to run?

I have resolved the conflicts and styling issues
